### PR TITLE
Implement follow-up attacks and update virtue stone weapons to use follow-up attacks

### PR DESCRIPTION
**Pull request by Dukilles**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Based on retail research, makes some adjustments to attackround and implements follow-up attacks. Update virtue stone weapons to use follow-up attacks.

### H2H attack round
- Using [Cichol's Mantle](https://www.ffxiah.com/item/26246/cichols-mantle), we were able to prove that each fist is rolled one after the other, with double/triple attacks following, but only the animation is changed to make the second swing a right hand.
- [Video](https://www.youtube.com/watch?v=mBUqbDieX1E)
- Left/Right/Left (DOUBLE, DOUBLE, NORMAL)
   - Screenshot proved too difficult to get here, but at ~43 seconds in you see 213, 214, 180
- Left/Right/Left (NORMAL, DOUBLE, DOUBLE)
   - image
- Left/Right/Left/Left (DOUBLE, DOUBLE, DOUBLE, DOUBLE)
   - image
- My solution to this is to `CreateAttacks` with the left hand twice, and force the animation of the second swing to `RIGHTATTACK` in `AddAttackSwings`
- I must admit that one out of every ~100 attack rounds will do something I can't explain like 4 consecutive right hands. I don't have an answer for that outlier.
- Mob's H2H direction on retail seems to be random, but for this PR they will follow players as they did before.

### Iga mod
- Iga mod adds to double attack rate for the offhand attack.
- If I need to remove this as out of scope, I certainly understand. Using `!isMainHand` while keeping the mod check where it was would admittedly be more in scope for this PR.

### Follow-up attacks
- Virtue Stones create follow-up attacks. Per retail testing, [this bg page](https://www.bg-wiki.com/ffxi/Category:Multi-Attack) seems to be the most accurate.
- After the standard attack round is created, `ProcFollowUpAttacks` will iterate through the swings and check for eligibility to store a follow-up swing. If a swing is eligible to proc a follow-up and does so successfully, it will be stored in order. Once all of the attack swings have been checked, follow-up attacks will be appended to the attack round in the order that they triggered.
- Normal, Double, Triple, Quad, and Samba type attacks are all eligible to proc virtue stone follow-ups. I am unsure about the others but added a method of type checking.
- Virtue Stones will be consumed as soon as they proc and are stored, even if the swing doesnt fire. (such as mob dying early in the attack round)
- I tried to do this in a way that would create some scaffolding for Dynamis [D] and Raetic weapons to be added in the future as follow-up attacks.

### [Faith Baghnakhs - H2H](https://ffxiclopedia.fandom.com/wiki/Faith_Baghnakhs)
- `!additem 5410 99` -- Virtue Stone Pouch
- `!additem 18360` -- Faith Baghnakhs
- standard h2h attack is two normal swings: Left/Right
- each fist will roll da/ta/qa and be added as single swings.
- Kicks proc next
- Each hand can use a stone and proc an additional follow-up attack.

#### Example Key
- X = standard, first two swings of the H2H attack round (can proc da/ta)
- l = additional swings from double/triple attack
- K = kick attack
- L = Follow-up Left
- R = Follow-up Right

#### Attack Order
- The structure looks like this: X l K L R ( L R can be R L, it depends on the order it was triggered)

#### No stones
- X (No procs)
- Xl (One fist double)
- Xll (One fist triple OR both fists double)
- Xlll (One fist double, one fist triple)
- Xllll (both fists triple)

#### After this, kick attacks will be added if they trigger
- XK (No procs)
- XlK (One fist double)
- XllK (One fist triple OR both fists double)
- XlllK (One fist double, one fist triple)
- XllllK (both fists triple)

#### Next, all attacks from the knuckles will try to proc a follow-up
- Follow-ups will be added in the order that they proc. Max one per hand.
- XKLR
- XlKLR OR XlKRL
- XllKLR OR XllKRL
- XlllKLR OR XlllKRL
- XllllKL OR XllllKR (caps at 8, although both stones will be consumed)

#### Example Key
- `!additem 18222`
- R - Right handed swing (2 Handed weapons are all `RIGHTATTACK`)

#### Examples
- Stone can proc on base attack or the da, or both ta
- R (no proc)
- RR (da)
- RRR (ta)
- Follow up will try to proc on all of these attacks and append to the end
- R(R) (no proc + stone)
- RR(R) (da + stone)
- RRR(R) (ta + stone)
#### Example Key
- `!additem 17710`
- you will need another sword to test the opposite hand
- R - Right handed swing
- L - Left handed swing
#### Examples
- Stone proc will add a follow-up attack to the same hand and append to the end of the round
- For the attacks below, assume that Justice Sword is in the right hand.
- R L
- R LL (left hand da)
- RR L (right hand da)
- RR LL (both hands da)
- if any of the R swings proc a stone, a R will be appended to the end:
- R L (R) (no proc + stone)
- R LL (R) (left hand da + stone)
- RR L (R) (right hand da + stone)
- RR LL (R) (both hands da + stone)
- If the Justice Sword is in the left hand, it will append to the end like this
- R L
- R LL (left hand da)
- RR L (right hand da)
- RR LL (both hands da)
- if any of the L swings proc a stone, a L will be appended to the end:
- R L (L) (no proc + stone)
- R LL (L) (left hand da + stone)
- RR L (L) (right hand da + stone)
- RR LL (L) (both hands da + stone)

### Final Notes
- Special thanks to @siknoz for helping with the research
- I am not sure if the indexes need to be adjusted in `gear_sets.lua`
- Please feel free to ask for any research. I have not yet uploaded any of it but certainly could.

### CODE REVIEW

File: src/map/attackround.cpp [hunk 1]
```diff
@@ -51,27 +51,30 @@ CAttackRound::CAttackRound(CBattleEntity* attacker, CBattleEntity* defender)
...
-    // Build dual wield off hand weapon attacks.
-    if (IsH2H())
+    if (IsH2H()) // Build H2H attacks.
     {
-        // Build left hand H2H attacks.
         CreateAttacks(dynamic_cast<CItemWeapon*>(attacker->m_Weapons[SLOT_MAIN]), LEFTATTACK);
-
-        // Build kick attacks.
-        CreateKickAttacks();
+        CreateAttacks(dynamic_cast<CItemWeapon*>(attacker->m_Weapons[SLOT_MAIN]), LEFTATTACK);
```
Comment (zach2good) [inline src/map/attackround.cpp]: To passers-by: this was gross before these changes were made and `CreateAttacks` has null-checks built in. Do not learn from this code though, this isn't how to use `dynamic_cast`!

(you don't need to change anything... unless you're feeling brave)

File: src/map/attack.h [hunk 2]
```diff
@@ -39,7 +39,8 @@ enum class PHYSICAL_ATTACK_TYPE
     RAPID_SHOT = 6,
     SAMBA      = 7,
     QUAD       = 8,
-    DAKEN      = 9
+    DAKEN      = 9,
+    FOLLOWUP   = 10
```
Comment (zach2good) [inline src/map/attack.h]: Trailing comma here, please

**Review by zach2good — APPROVED** [review]
Looks very well researched and well laid out to me, thanks!

**Review by WinterSolstice8 — APPROVED** [review]
Good job, thanks for all the effort here and thanks to Siknoz as well

File: src/map/attackround.cpp (hunk 1 above)
Comment (Xaver-DaRed) [reply src/map/attackround.cpp]: EDIT: Meh, more reading needed on my end

**Review by Xaver-DaRed — APPROVED** [review]

File: src/map/attackround.cpp (hunk 1 above)
Comment (Dukilles) [reply src/map/attackround.cpp]: > To passers-by: this was gross before these changes were made and `CreateAttacks` has null-checks built in. Do not learn from this code though, this isn't how to use `dynamic_cast`!
> 
> (you don't need to change anything... unless you're feeling brave)

I was feeling brave this morning, hehe. Sorry about that :D

**Review by zach2good — APPROVED** [review]

**Comment by zach2good** [comment]
Nice work 👍

**Pull request merged**